### PR TITLE
Provide variant of distribute_variable of DofManager.

### PR DESCRIPTION
**Pull request by jwboth**
The new distribute_variable() method allows for distributing solution to subsystems (of multiphysics problems) to the full solution. Both states as well as iterates can be updated. This change is suitable for 

This PR builds on current PR "Fix bug when using multiple active Ad variables". 

The functionality is demonstrated in porepy_lab, see ad-related/two-phase-flow-ad-test.py

### CODE REVIEW

File: src/porepy/numerics/ad/equation_manager.py [hunk 1]
```diff
@@ -296,18 +299,20 @@ def to_ad(
             ad_vars = initAdArrays([state[ind] for ind in self._variable_dofs])
             self._ad = {var_id: ad for (var_id, ad) in zip(self._variable_ids, ad_vars)}
         else:
-            active_variable_ids = [v.id for v in active_variables]
-
-            ad_variable_ids = list(
-                set(self._variable_ids).intersection(active_variable_ids)
+            potential_active_variable_ids = [v.id for v in active_variables]
+            active_variable_ids = list(
```
Comment (keileg) [inline src/porepy/numerics/ad/equation_manager.py:303]: Is the formatting correct here?

File: src/porepy/numerics/ad/equation_manager.py [hunk 2]
```diff
@@ -121,16 +120,20 @@ def __init__(
         self._identify_discretizations()
 
     # TODO why is this called local? and not global?
-    def local_dofs(self, true_ad_variables: Optional[list] = None) -> np.ndarray:
-        if true_ad_variables is None:
+    def local_dofs(self, active_variables: Optional[list] = None) -> np.ndarray:
```
Comment (keileg) [inline src/porepy/numerics/ad/equation_manager.py:123]: 1) Please use List, not list (the latter may not be properly recognized, from what I know, it was not properly introduced before Py 3.9).
2) Specify type of list, I suppose `List[str]`

File: src/porepy/numerics/ad/equation_manager.py [hunk 3]
```diff
@@ -121,16 +120,20 @@ def __init__(
...
+        if active_variables is None:
             dofs = np.hstack([d for d in self._variable_dofs])
         else:
-            true_ad_variable_ids = [v.id for v in true_ad_variables]
-            assert all([i in self._variable_ids for i in true_ad_variable_ids])
-            ad_variable_local_ids = [
-                self._variable_ids.index(i) for i in true_ad_variable_ids
+            active_variable_ids = [v.id for v in active_variables]
```
Comment (keileg) [inline src/porepy/numerics/ad/equation_manager.py:127]: Comments? What happens here?

File: src/porepy/numerics/ad/equation_manager.py [hunk 4]
```diff
@@ -319,19 +324,21 @@ def to_ad(
         else:
             # FIXME: This needs explanations
             prev_iter_vals_list = [state[ind] for ind in self._prev_iter_dofs]
-            non_ad_variable_ids = list(set(self._variable_ids) - set(ad_variable_ids))
-            non_ad_variable_local_ids = [
-                self._variable_ids.index(i) for i in non_ad_variable_ids
+            inactive_variable_ids = list(
```
Comment (keileg) [inline src/porepy/numerics/ad/equation_manager.py:327]: .. the nede for explanation is still there.

File: src/porepy/numerics/mixed_dim/dof_manager.py [hunk 5]
```diff
@@ -198,42 +196,115 @@ def distribute_variable(
...
-        global_dofs = np.zeros(total_dofs)
-        global_dofs[dofs] = 1
-
-        # Projection matrix from space of all dofs to restricted list of var
-        num_dofs = self.num_dofs(var=var)  # type: ignore
-        projection = sps.coo_matrix(
-            (np.arange(num_dofs), (np.arange(num_dofs), dofs)),
-            shape=(num_dofs, total_dofs),
-        )
+    def distribute_restricted_variable(
```
Comment (keileg) [inline src/porepy/numerics/mixed_dim/dof_manager.py:202]: What does restricted mean? How is this different from distribute variable? Should be clear in function documentation.

File: src/porepy/numerics/mixed_dim/dof_manager.py [hunk 6]
```diff
@@ -198,42 +196,115 @@ def distribute_variable(
...
+                        else:
+                            data[pp.STATE][pp.ITERATE] = {}
+                            data[pp.STATE][pp.ITERATE][var_name] = vals
+
+                    else:
+                        # If no values exist, there is nothing to add to
+                        data[pp.STATE] = {}
+                        data[pp.STATE][pp.ITERATE] = {var_name: values[dof[vi] : dof[vi + 1]]}
+
+    def transform_dofs(self, dofs: np.ndarray, var: Optional[list]) -> np.ndarray:
```
Comment (keileg) [inline src/porepy/numerics/mixed_dim/dof_manager.py:277]: Is transform a good word here? Restrict?

File: src/porepy/numerics/mixed_dim/dof_manager.py [hunk 7]
```diff
@@ -198,42 +196,115 @@ def distribute_variable(
...
+        """Transforms dofs associated to full list of dofs to a restricted list of dofs.
+
+        It is implicitly assumed that for each variable in var all versions on all possible
```
Comment (keileg) [inline src/porepy/numerics/mixed_dim/dof_manager.py:280]: I don't understand the assumption.

File: src/porepy/numerics/mixed_dim/dof_manager.py [hunk 8]
```diff
@@ -198,42 +196,115 @@ def distribute_variable(
...
+                            data[pp.STATE][pp.ITERATE][var_name] = vals
+
+                    else:
+                        # If no values exist, there is nothing to add to
+                        data[pp.STATE] = {}
+                        data[pp.STATE][pp.ITERATE] = {var_name: values[dof[vi] : dof[vi + 1]]}
+
+    def transform_dofs(self, dofs: np.ndarray, var: Optional[list]) -> np.ndarray:
+        """Transforms dofs associated to full list of dofs to a restricted list of dofs.
+
```
Comment (keileg) [inline src/porepy/numerics/mixed_dim/dof_manager.py:279]: Please describe the output (standard documentation), and how it relates to the input.

Comment (keileg) [reply src/porepy/numerics/mixed_dim/dof_manager.py:279]: .. that is, I what does the method really do?

**Review by keileg — COMMENTED** [review]
Some comments, mainly regarding documentation. 

The need for the functions is not clear, but I think we should be generous on which functions to include for a while, and then clean up later.

**Comment by keileg** [comment]
@jwboth I think this is outdated, is it okay if I delete it?

**Comment by jwboth** [comment]
> @jwboth I think this is outdated, is it okay if I delete it?

Go for it.

**Pull request closed without merging**